lib/api/CInferenceModelDefinition.cc

```diff
@@ -0,0 +1,459 @@
+
+    //input
+    rapidjson::Value inputObject = writer.makeObject();
+    m_Input.addToDocument(inputObject, writer);
+    writer.addMember(JSON_INPUT_TAG, inputObject, doc);
+
+    // preprocessors
+    rapidjson::Value preprocessingArray = writer.makeArray();
+    for (const auto& encoding : m_Preprocessors) {
+        rapidjson::Value encodingValue = writer.makeObject();
+        encoding->addToDocument(encodingValue, writer);
+        rapidjson::Value encodingEnclosingObject = writer.makeObject();
+        writer.addMember(encoding->typeString(), encodingValue, encodingEnclosingObject);
+        preprocessingArray.PushBack(encodingEnclosingObject, writer.getRawAllocator());
+    }
+    writer.addMember(JSON_PREPROCESSORS_TAG, preprocessingArray, doc);
+
+    //trained_model
+    if (m_TrainedModel) {
+        rapidjson::Value trainedModelValue = writer.makeObject();
+        m_TrainedModel->addToDocument(trainedModelValue, writer);
+        writer.addMember(JSON_TRAINED_MODEL_TAG, trainedModelValue, doc);
+    } else {
+        LOG_ERROR(<< "Trained model is not initialized");
+    }
+    writer.write(doc);
+    return stringBuffer.GetString();
+}
+
+void CTrainedModel::addToDocument(rapidjson::Value& parentObject,
+                                  TRapidJsonWriter& writer) const {
+    addJsonArray(JSON_FEATURE_NAMES_TAG, m_FeatureNames, parentObject, writer);
+
+    if (m_ClassificationLabels) {
+        rapidjson::Value classificationLabelsArray =
+            writer.makeArray(m_ClassificationLabels.get().size());
+        for (const auto& classificationLabel : *m_ClassificationLabels) {
+            rapidjson::Value classificationLabelValue;
+            classificationLabelValue.SetString(classificationLabel,
+                                               writer.getRawAllocator());
+            classificationLabelsArray.PushBack(classificationLabelValue,
+                                               writer.getRawAllocator());
+        }
+        writer.addMember(JSON_CLASSIFICATION_LABELS_TAG, classificationLabelsArray, parentObject);
+    }
+
+    switch (m_TargetType) {
+    case E_Classification:
+        writer.addMember(JSON_TARGET_TYPE_TAG, JSON_TARGET_TYPE_CLASSIFICATION, parentObject);
+        break;
+    case E_Regression:
+        writer.addMember(JSON_TARGET_TYPE_TAG, JSON_TARGET_TYPE_REGRESSION, parentObject);
+        break;
+    }
+}
+
+const CTrainedModel::TStringVec& CTrainedModel::featureNames() const {
+    return m_FeatureNames;
+}
+
+void CTrainedModel::featureNames(const CTrainedModel::TStringVec& featureNames) {
+    m_FeatureNames = featureNames;
+}
+
+void CTrainedModel::targetType(CTrainedModel::ETargetType targetType) {
+    m_TargetType = targetType;
+}
+
+CTrainedModel::ETargetType CTrainedModel::targetType() const {
+    return m_TargetType;
+}
+
+void CInferenceModelDefinition::fieldNames(const std::vector<std::string>& fieldNames) {
```

Comment:
You have the typedef:
```suggestion
void CInferenceModelDefinition::fieldNames(const TStrVec& fieldNames) {
```

Reply:
done